File: countdownlauhging/src/main/java/com/laughing/android/countdownlauhging/MainActivity.java
package com.laughing.android.countdownlauhging;

import android.os.Bundle;
import android.os.CountDownTimer;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {

    private Button btn_djs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        btn_djs = (Button) findViewById(R.id.btn_djs);
        //new倒计时对象,总共的时间,每隔多少秒更新一次时间
        final MyCountDownTimer myCountDownTimer = new MyCountDownTimer(60000,1000);

        //给Button设置点击时间,触发倒计时
        btn_djs.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                myCountDownTimer.start();
            }
        });
    }

    //复写倒计时
    private class MyCountDownTimer extends CountDownTimer {

        public MyCountDownTimer(long millisInFuture, long countDownInterval) {
            super(millisInFuture, countDownInterval);
        }

        //计时过程
        @Override
        public void onTick(long l) {
            //防止计时过程中重复点击
            btn_djs.setClickable(false);
            btn_djs.setText(l/1000+"s");

        }

        //计时完毕的方法
        @Override
        public void onFinish() {
            //重新给Button设置文字
            btn_djs.setText("重新获取验证码");
            //设置可点击
            btn_djs.setClickable(true);
        }
    }
}

File: kotlin/src/main/java/com/laughing/kotlin/kotlin/MainActivity.kt
package com.laughing.kotlin.kotlin

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.Toast
import com.laughing.kotlin.R
import com.laughing.kotlin.kotlin.bean.Person
import com.laughing.kotlin.kotlin.bean.Student
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    companion object {
        val NAME_KEY = "MainActivity::key"
        val AGE_KEY = "MainActivity::value"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initView()
        initData()
        getIntentVaule()
        initListener()
    }

    /**
     * 获取intent中从上一个页面传过来的值
     */
    private fun getIntentVaule() {
        val name = intent.getStringExtra(NAME_KEY)
        val age = intent.getIntExtra(AGE_KEY, 0)
        Log.e("TAG", "$name  $age")

    }

    private fun initView() {
        tv_name.setText("Laughing")
        var p = Person("zhang san", 18)
        var s = Student("zhang san", 10, courseCount = 3)
        var s1 = Student("zhang san", 18, 3)
        var s2 = Student("zhang san", 18, 3, 4.0)
        Log.e("TAG", "${s.name}")
        Log.e("TAG", "${s.height}")
        Log.e("TAG", s.name + "  " + s.age + "  " + s.height)

//        var s3: Student2? = null
//        s3?.play()

        var s3 = Student("li si", 10, 3)
        s3.height = 1.50
        Log.e("TAG", "${s3.height} ------------------------>")
        var height = s3.height
        Log.e("TAG", "$height ----------------height-------->")


        val linearLayout = findViewById<LinearLayout>(R.id.parent)
        val imageView = linearLayout.get(2) as ImageView
        imageView.setImageResource(R.mipmap.ic_launcher)
    }


    private fun initData() {
        //Kotlin扩展类用法
//        var vg: ViewGroup? = null
//        var childAt = vg?.get(3)


    }

    fun ViewGroup.get(index: Int) = this.getChildAt(index)

    private fun initListener() {
        tv_name.setOnClickListener { v ->
            Toast.makeText(this, "hello", Toast.LENGTH_SHORT).show()
            println("hello")

        }
    }
}


